doc/js-spec/oic.md

```diff
@@ -0,0 +1,480 @@
+
+The following error event is used for protocol errors:
+```javascript
+interface OicErrorEvent: Event {
+  OicError error;
+};
+```
+
+Code Examples
+-------------
+### Getting device configuration
+
+```javascript
+var oic = require('oic');
+if (oic.device.uuid) {  // configuration is valid
+  startServer();
+  startClient();
+} else {
+  console.log("Error: device is not configured.");
+}
+```
+
+### OIC Client contolling a remote red LED.
+```javascript
+// Discover a remote red light, start observing it, and make sure it's not too bright.
+var red = null;
+function startClient() {
+  // discover resources
+  oic.onresourcefound = function(event) {
+    if(event.resource && event.resource.id.path == "/light/ambience/red") {
+      red = event.resource;
+      red.addEventListener('update', redHandler);
+    }
+  }
+  oic.findResources({ resourceType: “oic.r.light” })
+    .then( () => { console.log("Resource discovery started.");})
+    .catch((e) => {
+      console.log("Error finding resources: " + e.message);
+    });
+};
+
+function redHandler(red, updates) {
+  console.log("Update received on " + red.id);
+  console.lof("Running local business logic to determine further actions...");
+  if (red && red.properties.dimmer > 0.5) {
+    // do something, e.g. limit output
+    oic.update({ id: red.id, properties.dimmer: 0.5 })
+      .then(() => { console.log("Changed red light dimmer"); })
+      .catch((e) => { console.log("Error changing red light"); });
+  }
+};
+```
+
+### OIC Server exposing a local blue LED.
+
+```javascript
+var lightResource = null;
+function startServer() {
+  // register the specific resources handled by this solution
+  // which are not exposed by the device firmware
+  oic.registerResource({
+    id: { deviceId: oic.device.uuid; path: "/light/ambience/blue" },
+    resourceTypes: [ "light" ],
+    interfaces: [ "/oic/if/rw" ],
+    discoverable: true,
+    observable: true,
+    properties: { color: "blue", dimmer: 0.2 }
+  }).then((res) => {
+    console.log("Local resource " + res.id.path + " has been registered.");
+    lightResource = res;
+    oic.addEventListener("updaterequest", onLightUpdate);
+    oic.addEventListener("observerequest", onLightObserve);
+    oic.addEventListener("deleterequest", onLightDelete);
+    oic.addEventListener("retrieverequest", onLightRetrieve);
+    oic.addEventListener("createrequest", onLightCreate);
+    }
+  }).catch((error) => {
+    console.log("Error creating resource " + error.resource.id.path + " : " + error.message);
+  });
+};
+
+function onLightRetrieve(event) {
+  if (event.target.id.path == lightResource.id.path) {
```

Comment:
Are these paths different types? If not it should be three equals "===", otherwise it will get warnings